File: 1475.final-prices-with-a-special-discount-in-a-shop/solution.kt
class Solution {
    fun finalPrices(prices: IntArray): IntArray {
        return prices.mapIndexed { index, price -> price - discount(prices, index) }.toIntArray()
    }

    private fun discount(prices: IntArray, index: Int): Int {
        for (j in index + 1 .. prices.lastIndex) {
            if (prices[j] <= prices[index]) {
                return prices[j]
            }
        }
        return 0
    }
}

File: 1475.final-prices-with-a-special-discount-in-a-shop/solution.go
func finalPrices(prices []int) []int {
    res := make([]int, len(prices))
    copy(res, prices)

    var stack []int

    for i, price := range prices {
        for len(stack) > 0 && prices[stack[len(stack) - 1]] >= price {
            res[stack[len(stack) - 1]] -= price
            stack = stack[:len(stack) - 1]
        }    
        stack = append(stack, i)
    }
    return res
}